monitor/internal/pod/delete_controller.go

```diff
@@ -95,38 +119,100 @@ func deleteControllerProcessItem(backgroundCtx context.Context, c client.Client,
+			zap.L().Warn("DeleteController: Failed to get pod from Kubernetes API", zap.String("puID", podUID), zap.String("namespacedName", req.String()), zap.Error(err))
 		}
 		return
 	}
 
 	// the edge case: a pod with the same namespaced name came up and we have missed a delete event
 	// this means that this pod belongs to a different PU and must live, therefore we try to delete the old one
-	if nativeID != string(pod.GetUID()) {
-		zap.L().Error("Pod does not have expected native ID, we must have missed an event and the same pod was recreated. Trying to destroy PU", zap.String("puID", nativeID), zap.String("namespacedName", req.String()), zap.String("podUID", string(pod.GetUID())))
+
+	// the following code also takes care of any restarts in the Pod, the restarts can be caused by either
+	// the sandbox getting killed or all the containers restarting due a crash or kill.
+
+	// Now destroy the PU only if the following
+	// 1. Simple case if the pod UID don't match then go ahead and destroy the PU.
+	// 2. When the pod UID match then do the following:
+	//		2.a Get the current SandboxID from the pod.
+	// 		2.b Get the sandboxID from the map.
+	// 		2.c If the sandBoxID differ then send the destroy event for the old(map) sandBoxID.
+
+	// 1st case, simple if the pod UID don't match then just call the destroy PU event and delete the map entry with the old key.
+	if string(pod.UID) != delObj.podUID {
+
+		zap.L().Warn("DeleteController: Pod does not have expected native ID, we must have missed an event and the same pod was recreated. Trying to destroy PU", zap.String("puID", podUID), zap.String("namespacedName", req.String()), zap.String("podUID", string(pod.GetUID())))
 		if err := pc.Policy.HandlePUEvent(
 			ctx,
-			nativeID,
+			podUID,
 			common.EventDestroy,
 			policy.NewPURuntimeWithDefaults(),
 		); err != nil {
 			// we don't really care, we just warn
-			zap.L().Warn("failed to handle destroy event", zap.String("puID", nativeID), zap.String("namespacedName", req.String()), zap.Error(err))
+			zap.L().Warn("DeleteController: Failed to handle destroy event", zap.String("puID", podUID), zap.String("namespacedName", req.String()), zap.Error(err))
 		}
 		// we only fire events away, we don't really care about the error anyway
 		// it is up to the policy engine to make sense of that
-		delete(*m, nativeID)
+		delete(m, podUID)
+		return
+	}
+
+	// now the 2nd case, when pod UID match
+	if string(pod.UID) == delObj.podUID {
+		zap.L().Debug("DeleteController: the pod UID Match happened, delete the", zap.String("podName:", req.String()), zap.String("podUID", string(pod.UID)))
+		// 2a get the current sandboxID
+		if sandboxExtractor == nil {
+			return
+		}
+		currentSandboxID, err := sandboxExtractor(ctx, pod)
+		if err != nil {
+			zap.L().Error("DeleteController: cannot extract the SandboxID, return", zap.String("namespacedName", req.String()), zap.String("podUID", string(pod.GetUID())))
```

Comment:
one last log change: again this can happen regularly, so let's make this debug please